README.md
# 阿里云 OSS 支持插件 (Aliyun OSS For WordPress)

本插件主要为 Wordpress 提供基于阿里云 OSS 的远程附件存储功能，并且最大限度的依赖 Wordpress 本身功能扩展来实现，以保证插件停用或博客搬迁时可以快速切换回原来的方式。

## 插件特色

1. 支持 Aliyun OSS 的图片服务（根据参数获得不同尺寸的图片）
2. 自定义文件在 Bucket 上的存储位置  
3. 支持 Https 站点
4. 支持阿里云内网和 VPC 网络
6. 全格式附件支持，不仅仅是图片
7. 支持 WordPress 4.4+ 新功能 srcset，在不同分辨率设备上加载不同大小图片
8. 支持 WordPress 5.0+ 默认古藤堡编辑器Gutenburg
9. 支持阿里云url鉴权，实现高级别防盗链功能，支持阿里云CDN的A、B、C所有方式鉴权，推荐A方式
10. 支持在 WordPress 后台编辑图片
11. 支持预设图片样式，图片保护，自定义分割符
12. 中英文双语支持，方便使用英文为默认语言的同学
13. 支持在其他插件/主题中通过系统钩子调用插件功能
14. 代码遵循 PSR-4 规则编写

## 插件使用

关于插件使用方式的 Wiki: [Quick start](https://github.com/IvanChou/aliyun-oss-support/wiki/Quick-start)

### 下载

[latest release](../../releases/latest)

### 安装

将插件解压上传到 `/wp-content/plugins/` 或者通过 WordPress 插件中心上传安装

### 配置

启用插件 `Aliyun OSS`

进入设置页面 完成相关设置

- 如果你使用的 ECS 与 OSS 在同一区域，可以开启『**内网/internal**』选项，节约流量
- 『Bucket 域名/Bucket Host』一项会自动补全，也可以手动设置为你的 **自定义域名** 或 CDN 域名

![screenshot](screenshot.png)

## 启用 OSS 图片服务

阿里云 OSS 提供了根据 url 参数来获得各种尺寸的 `阿里云OSS图片处理服务（Image Service，简称 IMG）`, 相比起 WordPress 上传的时候生成各种尺寸的图片, 这是一种更优雅的解决方案, 占用的存储空间更小, 尺寸变更更灵活。

如何开启并配置图片服务, 请参见: [How to use Image Service](https://github.com/IvanChou/aliyun-oss-support/wiki/How-to-use-Image-Service)

另外还有几点需要你了解:

1. 开启图片服务时, 只有原图会被上传到 OSS, 缩略图本地依旧会生成但不会上传
2. 基于第 1 条, 建议开启图片服务后就不要关了, 关掉会导致之前上传的图片缩略图无法访问（文章中的图片不受影响）
3. 基于第 2 条, 如果你确实想关掉图片服务，参见下一项里面的解决方案

## 启用 CDN URL鉴权

1.阿里云URL鉴权功能由CDN服务提供，安全级别比较高，启用URL鉴权功能时，需要先在阿里云CDN控制面板-域名管理-管理-访问控制中设置鉴权方式和鉴权主/备Key，Key为完全自定义，请勿使用您的密码/AK/SK，鉴权方式推荐设置为A方式。具体设置详见[什么是url防盗链](https://help.aliyun.com/document_detail/85117.html?spm=5176.11785003.domainDetail.2.1192142f7p1JWz)

2.在阿里云控制台中设置好鉴权方式和Key之后，打开插件URL鉴权功能，填入您的鉴权方式和Key，以及您希望URL失效的时间（按小时计），即可实现URL鉴权

3.如果不希望开启URL鉴权，阿里云CDN和OSS还带有更为简单的Refer黑/白名单防盗链功能，安全性稍差且不带链接失效时间设置，详见[CDN如何配置refer防盗链](https://help.aliyun.com/document_detail/27134.html?spm=5176.11785003.domainDetail.1.1192142f7p1JWz)[OSS防盗链使用指南](https://help.aliyun.com/document_detail/31869.html?spm=5176.8466029.referer.1.621714504i4WC3)

## 启用插件后，老文件无法访问

参见 WIKI：[How to handle old images](https://github.com/IvanChou/aliyun-oss-support/wiki/How-to-handle-old-images)

## 关于不在本地服务器上保留文件

『不在本地服务器上保留文件』建议不要开启, 理由如下:

1. 如果同时开启『图片服务』, 当你想停用这个插件的时候不可避免的会遇到缩略图丢失问题
2. 如果没有同时开启『图片服务』, 当你从后台删除图片或附件时, OSS 里面的缩略图无法被删掉

****

## 题外

本插架由官方商店中 马文建(@mawenjian) 同学的[「阿里云附件」](https://github.com/mawenjian/aliyun-oss-support)插件拓展而来。由于马同学在曾经的某段时间里没能即时维护这个项目，也没有开源，于是我在修复 bug 并 rebuild 后，将这个野生的修订版发布到阿里云社区，意外获得了 ACE 社区官方管理组的推荐。

后来，马同学 release 了 2.0 版本并开源他的项目了，~~我就中止了这边的维护~~。但依旧是有网友提 Issue 或发邮件来询问，加上自己的需求，有时间的时候，也就修补一下大家反应的问题，也许还是会有人会用到。

由于插件沿用了马同学插件的名字，并 WordPress 官方不再允许在未经授权的情况下使用知名商标（如：Aliyun) 作为插件名称的一部分，所以这个插件并没有提交官方商店的计划。（重新想个名字对我来说太麻烦了~~(￣▽￣)）

## 更新日志

[CHANGELOG.md](CHANGELOG.md)

## 项目依赖

- https://github.com/aliyun/aliyun-oss-php-sdk
- 
## 贡献代码

1. Fork 这个仓库
2. Clone 源码
3. 执行 `composer install`
4. 安装到本地 WordPress 中或者 执行 `docker-compose up` 启用一个新实例
5. 完成你的修改并测试
6. 提交一个 Pull Request

## 开源协议

BSD



CHANGELOG.md
## ChangeLog - 阿里云附件存储
____________________

### 版本号: 3.4.1

修正日期：2022-09-25

1. 移除图片服务「自定义分隔符」中默认的 `@` #107

修正日期：2022-09-25

1. 升级 aliyun-oss-php-sdk
2. 移除 plugin-update-checker
3. 移除对 PHPMailer 的依赖

### 版本号: 3.3.0

修正日期：2021-11-24

1. 支持 Aliyun URL鉴权
2. 移除 repo 中的 vendor 目录，使用 composer 更新依赖
3. 兼容 Wordpress 5.4+
4. 增加 docker-compose.yml 方便测试
5. 新增 乌兰察布、河源、广州、成都、英国 大区 #124

### 版本号: 3.2.7

修正日期：2019-04-01

1. 新增『禁用主动上传文件到 OSS』 选项
2. 升级 plugin-update-checker 到 4.5.1

### 版本号: 3.2.6

修正日期：2019-03-27

1. 升级 aliyun-oss-php-sdk 到 2.3.0
2. 新增异常处理
3. 优化加载机制，尽可能的减小对系统资源的消耗（虽然本来就不占资源）

### 版本号: 3.2.5

1. 修正: 服务器本地不保留文件时，会导致 OSS 上对应的文件删不掉.
    若不開啟『图片服务』, 所生成的縮略圖仍然不會刪除.
2. 修正: 服务器本地不保留文件时，会导致无法安装插件.
3. 修正: 中文文件名會導致圖片服務連結錯誤.

### 版本号: 3.2.4

修正日期：2018-11-07

#### 修订内容：

1. 为 Exclude 项追加 Regexp Formatter, 避免不正确的表达式导致 Warning
2. 添加 Exclude 选项的翻译

### 版本号: 3.2.3

修正日期：2018-11-06

#### 修订内容：

1. 修复样式文件下载白屏 BUG #79
2. 新增 Exclude 模式，可以排除指定目录 #61 #80

### 版本号: 3.2.2

修正日期：2018-08-21

#### 修订内容：

1. 修复不能使用 OSS 根目录作为存储路径的 bug #70
2. 修复图片服务开启时, 缩略图依旧被上传的 bug #72
3. 新增插件自动更新功能
4. 为 Bucket Host 增加 fallback 值
5. 调整部分配置选项

### 版本号: 3.2.1

修正日期：2018-07-31

#### 修订内容：

1. 更换了缩略图、编辑图的上传机制, 使兼容范围更广
2. 修复了 OSS 上文件不能删除的 BUG
3. 修复设置页面的一处 BUG
4. 更新数据节点

### 版本号: 3.1.7

修正日期：2018-07-25

#### 修订内容：

1. 新增阿里云图片样式导出功能, 再也不用去阿里云后台一项一项的设置了
2. 修复了更换主题后, 在后台编辑之前的特色图片时可能会失败的 bug

### 版本号: 3.1.6

修正日期：2018-02-04

#### 修订内容：

1. 升级 aliyun-oss-php-sdk 到 2.3.0
2. 新增「图片保护」和「自定义分割符」选项
3. 增加三个自定义 filter/action, 方便在其他地方调用插件功能 

### 版本号: 3.1.5

修正日期：2017-10-11

#### 修订内容：

1. 启用「预设图片样式」情况下, 由于 OSS 服务的 bug, 不对 gif 做处理. [#43](/../../issues/43)

### 版本号: 3.1.4

修正日期：2017-03-30

#### 修订内容：

1. 支持 PHP 5.3+ (以前是 5.4+)

### 版本号: 3.1.3

修正日期：2017-03-26

#### 修订内容：

1. 新增 OSS 节点及翻译
2. 新增 VPC 网络支持

### 版本号: 3.1.2

修正日期：2017-02-12

#### 修订内容：

1. 修复启用 图片预设样式 时，原图无法显示的 bug

### 版本号: 3.1.1

修正日期：2017-01-14

#### 修订内容：

1. 不再使用 phar 打包

### 版本号: 3.1.0

修正日期：2017-01-09

#### 修订内容：

1. Aliyun OSS 图片服务规则升级,移除 img_host 参数
2. 追加一个隐藏参数: keep_settings -> 卸载不删除配置

### 版本号: 3.0.4

修正日期：2016-12-22

#### 修订内容：

1. 修正本地不保留文件时,同名文件覆盖. [#20](/../../issues/20)
2. 升级阿里云 OSS SDK 到 2.2.1
3. 添加自动打包脚本

### 版本号: 3.0.3

修正日期：2016-11-16

#### 修订内容：

1. 修复不保存本地文件条件下, medium_large 和 large 尺寸一样时上传 OSS 失败的 Bug. [#18](/../../issues/18) [#23](/../../issues/23)
2. 新增日本节点、法兰克福节点的支持

### 版本号：3.0.2

修正日期：2016-10-09

#### 修订内容：

1. 修复 [ISSUE #18](/../../issues/18)
____________________

### 版本号：3.0.1

修正日期：2016-10-06

#### 修订内容：

1. 修复 basename() 处理中文开头的文件名的 bug

____________________

### 版本号：3.0.0

修正日期：2016-10-03

#### 修订内容：

1. Fix a bug in image resize rule

____________________

### 版本号：3.0.0-beta2

修正日期：2016-09-24

#### 修订内容：

1. Fix bug [#13](/../../issues/13) 
____________________

### 版本号：3.0.0-beta

修正日期：2016-09-17

#### 修订内容：

1. Add new feature -- support watermark [#15](/../../issues/15) 

2. Add new feature -- save edited image [#13](/../../issues/13) 

3. Rebuild all codes under PSR-4 and use phar file to release.

4. Update Aliyun Oss SDK to the latest version.

5. Settings page supports both English and 中文.

____________________

### 版本号：2.4.1

修正日期：2016-07-12

#### 修订内容：

1. Fix the bug [#ISSUE 11](/../../issues/10) 

2. 启用图片服务时，不再需要预设图片处理方式，改为按 Wordpress 提供的尺寸自动处理

3. 修复上一个版本引入的一个语法错误

____________________

### 版本号：2.4.0

修正日期：2016-04-30

#### 修订内容：

1. 增加启用图片服务时，对 srcset 的支持（ 支持 wordpress 4.4+ ）

2. 修改数据节点的链接，原链接失效了

##### 感谢
[#9](/../../issues/9) （感谢 @hairui219 提供）

____________________

### 版本号：2.3.2

修正日期：2015-11-14

#### 修订内容：

1. 更新 SDK (2015-08-19 版)

2. 添加对数据节点设置的选项，考虑到以后可能还有各种节点，干脆留了个框自己填

3. 重新封装原来的 Aliyun_oss 对象，使其在使用上和 ACE 引擎内置的 Alibaba::Storge 类保持一致，现在 ACE 和其他环境通用一个版本 

4. 使用单例，之前脑子进屎了，没上传一个对象就要 new 一个对象，简直是浪费（可能是因为自己不需要上传缩略图，所以当时没 care，建议大家也使用 OSS 的图片服务，很 nice 的其实）

5. 解决设置本地不保存文件时，安装主题、插件失败，并被上传到 OSS 的 BUG

##### 感谢各位的提的 Issue，由于已经转到 ruby 开发了，php 很少接触了，不能保持维护，实在对不住

____________________

### 版本号：2.1.3

修正日期：2015-01-19

#### 修订内容：

1. 将设置界面的 Secret Key 隐藏，Secret Key 不再会被读取到客户端

2. 处理中文乱码问题

特别注意：处理中文问题的时候，我发现好像是 Aliyun SDK 的问题，细看了下源码，这个 SDK 写得有点怨念，对 Ali 的好感一下子掉了好多

具体修改了 sdk.class.php line 1290 ~ 1291

准备反馈给官方~

##### 感谢：网友 sjw(at)cnsjw.cn 的建议 和 风的涟漪协助测试

____________________

### 版本号：2.1.2

修正日期：2014-12-26

#### 修订内容：

修正在非当前月份文章中上传图片时，由于缩略图无法上传导致错误的 bug.

##### 感谢：网友 sjw(at)cnsjw.cn 协助修订

____________________

### 版本号：2.1.1

修正日期：2014-12-25

#### 修订内容：

处理 OSS 自定义存储 path 为空时出现的多斜线 Bug

##### 感谢：网友 sjw(at)cnsjw.cn 反馈

____________________

### 版本号：2.1

修正日期：2014-11-7

#### 修订内容：

在某些环境中，启用插件时提示

```
这个插件启用过程中产生了3个字符的**异常输出**。如果您遇到了……
```

原因为 sdk.class.php 这个文件编码问题，已修复

若仍有这个报错，请尝试转换文件编码与您的网站环境编码一致。

##### 感谢：网友 tang6818(at)foxmail.com 反馈

____________________

### 版本号：2.0

修正日期：2014-9-1

#### 修订内容：
1. 完全重构，优化代码
2. 支持 Aliyun OSS 的图片服务
3. 改变钩子嵌入机制，支持所有附件（以前的版本只上传图片，而且启用时其他附件完全不可用了，坑！！）
4. 添加卸载、不残留
5. 支持 Aliyun ACE （ 切换到 ACE 分支 ）


____________________

### 版本号：1.1

修正日期：2014-8-27

#### 修订项目：
1. 插件年久失修，其内部调用的 Aliyun OSS php SDK 已升级
2. WordPress 3.5以后 设置->多媒体 中没有路径配置，导致配置不便
3. Aliyun OSS 可以绑定自己的域名，插件中不能简单的设置

#### 修订内容：
1. 升级 Aliyun-OSS-SDK 到 1.1.6 版本 (2014-06-25更新)
2. 设置中可直接配置访问路径 Url，支持已绑定到 OSS 的独立域名
3. 支持自定义 OSS 上文件的存放目录 （不影响本地存储，中途若修改请手动移动 OSS 上文件，否则可能链接不到之前的资源）
4. 修正原插件 bug 若干
5. 优化代码 ~~（移除所有 Notice 级报错）~~


oss-support.php
<?php
/**
 * Plugin Name: Aliyun Storage Support
 * Plugin URI: "https://yii.im/post/aliyun-oss-support-plugin-for-wordpress"
 * Description: 使用阿里云存储OSS作为附件存储空间。This is a plugin that used Aliyun Cloud Storage(Aliyun OSS) for attachments remote saving.
 * Author: ichou(Ivan Chou)
 * Author URI: https://yii.im/
 * Version: 2.5.0
 * Updated_at: 2016-07-24
 */

/*  Copyright 2016  ichou  (email : yiichou@gmail.com)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

require(dirname(__FILE__).DIRECTORY_SEPARATOR.'SDK/alioss.class.php');

//  plugin url
define('OSS_BASEFOLDER', plugin_basename(dirname(__FILE__)));

$oss_options = get_option('oss_options');

/**
 * 初始化
 */
function oss_set_options() {
    $options         = array(
        'bucket'     => "",
        'ak'         => "",
        'sk'         => "",
        'end_point'  => "",
        'path'       => "",
        'static_url' => "",
        'img_url'    => ""
    );
    add_option('oss_options', $options, '', 'yes');
}
register_activation_hook(__FILE__, 'oss_set_options');

/**
 * 设置提醒
 *
 * @param $oss_options
 */
function oss_admin_warnings($oss_options) {

    $oss_bucket = isset($oss_options['bucket']) ? esc_attr($oss_options['bucket']) : null;
	if ( !$oss_bucket && !isset($_POST['submit']) ) {
		function oss_warning() {
			echo "<div id='oss-warning' class='updated fade'><p><strong>".__('OSS is almost ready.')."</strong> ".sprintf(__('You must <a href="%1$s">enter your OSS Bucket </a> for it to work.'), "options-general.php?page=" . OSS_BASEFOLDER . "/oss-support.php")."</p></div>";
		}
		add_action('admin_notices', 'oss_warning');
		return;
	} 
}
oss_admin_warnings($oss_options);

/**
 * 上传原文件到 OSS (并根据设定清理本地文件)
 *
 * @param $file
 * @return mixed
 */
function upload_orign_2_oss($file)
{
    if ($_GET["action"] == 'upload-plugin' || $_GET["action"] == 'upload-theme') 
        return $file;

    $wp_uploads = wp_upload_dir();
    $oss_options = get_option('oss_options', TRUE);
    $config = array(
            'id'     => esc_attr($oss_options['ak']),
            'key'    => esc_attr($oss_options['sk']),
            'bucket' => esc_attr($oss_options['bucket']),
            'end_point' => esc_attr($oss_options['end_point'])
        );
    $oss_upload_path = trim($oss_options['path'],'/');
    $oss_nolocalsaving = (esc_attr($oss_options['nolocalsaving'])=='true') ? true : false;

    $object = str_replace($wp_uploads['basedir'], '', $file['file']);
    $object = ltrim($oss_upload_path . '/' .ltrim($object, '/'), '/');

    $aliyun_oss = Alibaba::Storage($config);

    $opt['Expires'] = 'access plus 1 years';
    $aliyun_oss->saveFile( $object, $file['file'], $opt);

    if($oss_nolocalsaving && false === strpos($file['type'], 'image')){
        _delete_local_file($file['file']);
    }

    return $file;

}
add_filter('wp_handle_upload', 'upload_orign_2_oss', 30);

/**
 * 上传生成的缩略图到 OSS (并根据设定清理本地文件)
 *
 * @param $metadata
 * @return mixed
 */
function upload_thumb_2_oss($metadata)
{
    if ( preg_match('/\d{4}\/\d{2}/', $metadata['file'], $m) ) {
        $wp_uploads = wp_upload_dir($m[0]);
    } else {
        return $metadata;
    }
    
    $oss_options = get_option('oss_options', TRUE);
    $config = array(
            'id'     => esc_attr($oss_options['ak']),
            'key'    => esc_attr($oss_options['sk']),
            'bucket' => esc_attr($oss_options['bucket']),
            'end_point' => esc_attr($oss_options['end_point'])
        );
    $oss_upload_path = trim($oss_options['path'],'/');
    $oss_nolocalsaving = (esc_attr($oss_options['nolocalsaving'])=='true') ? true : false;

    $opt['Expires'] = 'access plus 1 years';

    if (isset($metadata['sizes']) && count($metadata['sizes']) > 0) {
        foreach ($metadata['sizes'] as $val) {
            $object = ltrim($oss_upload_path . '/' .trim($wp_uploads['subdir'], '/').'/'.ltrim($val['file'], '/'), '/');
            $file = $wp_uploads['path'].'/'.$val['file'];

            if($oss_options['img_url'] == "") {
                $aliyun_oss = Alibaba::Storage($config);
                $aliyun_oss->saveFile( $object, $file, $opt);
            }

            if($oss_nolocalsaving)
                _delete_local_file($file);
        }
    }

    if($oss_nolocalsaving) {
        $file = $wp_uploads['basedir'].'/'.$metadata['file'];
        _delete_local_file($file);
    }

    return $metadata;
}
add_filter('wp_generate_attachment_metadata', 'upload_thumb_2_oss', 60);

/**
 * 删除远程服务器上的单个文件
 * 
 * @param $file
 * @return mixed
 */
function delete_remote_file($file)
{
    if(!false == strpos($file, '@'))
        return $file;

    $oss_options = get_option('oss_options', TRUE);
    $config = array(
            'id'     => esc_attr($oss_options['ak']),
            'key'    => esc_attr($oss_options['sk']),
            'bucket' => esc_attr($oss_options['bucket']),
            'end_point' => esc_attr($oss_options['end_point'])
        );
    $oss_upload_path = trim($oss_options['path'],'/');
    $wp_uploads = wp_upload_dir();

    $del_file = str_replace($wp_uploads['basedir'], '', $file);
    $del_file = ltrim($oss_upload_path . '/' .ltrim($del_file, '/'), '/');

    $aliyun_oss = Alibaba::Storage($config);

    $aliyun_oss->delete($del_file);

    return $file;
}
add_action('wp_delete_file', 'delete_remote_file');

/**
 * 删除本地的缩略图（修正由于启用图片服务导致的原生方法删除不了缩略图）
 * 仅在开启图片服务时启用
 * 
 * @param $file
 * @return mixed
 */
function delete_thumb_img($file)
{
    if(!false == strpos($file, '@'))
        return $file;

    $file_t = substr($file, 0, strrpos($file, '.'));
    array_map('_delete_local_file', glob($file_t.'-*'));
    return $file;
}

/**
 * 修改从数据库中取出的图片信息，以使用 Aliyun 的图片服务
 * 仅在开启图片服务时启用
 *
 * @param $data
 * @return mixed
 */
function modefiy_img_meta($data) {
    $filename = basename($data['file']);
    $fileformat = strrchr($filename,'.');
    if(isset($data['sizes'])) {
        foreach ($data['sizes'] as $size => $info) {
            $data['sizes'][$size]['file'] = "{$filename}@{$info['height']}h_{$info['width']}w_1l_1c{$fileformat}";
        }
    }

    return $data;
}

/**
 * 重置图片链接，使用独立的图片服务器。
 * 仅在开启图片服务时启用
 *
 * @param $url
 * @param $post_id
 * @return mixed
 */
function modefiy_img_url($url, $post_id) {
    $wp_uploads = wp_upload_dir();
    $oss_options = get_option('oss_options', TRUE);

    if(wp_attachment_is_image($post_id)){
        $img_baseurl = rtrim($oss_options['img_url'], '/');
        if(rtrim($oss_options['path'], '/') != ""){
            $img_baseurl = $img_baseurl .'/'. rtrim($oss_options['path'], '/');
        }
        $url = str_replace(rtrim($wp_uploads['baseurl'], '/'), $img_baseurl, $url);
    }
    return $url;
}

/**
 * 重置 srcset 图片链接，使用独立的图片服务器。
 * 仅在开启图片服务时启用
 *
 * @param $sources
 * @return mixed
 */
function modefiy_img_srcset_url($sources) {
    $oss_options = get_option('oss_options', TRUE);
    foreach( $sources as $w => $img )
        $sources[$w]['url'] = str_replace($oss_options['static_url'], $oss_options['img_url'], $img['url']);
    return $sources;
}

if(!$oss_options['img_url'] == ""){
    add_action('wp_delete_file', 'delete_thumb_img', 99);
    add_filter('wp_get_attachment_metadata', 'modefiy_img_meta', 990);
    add_filter('wp_calculate_image_srcset_meta', 'modefiy_img_meta', 990);
    add_filter('wp_get_attachment_url', 'modefiy_img_url', 30, 2);
    add_filter('wp_calculate_image_srcset', 'modefiy_img_srcset_url', 30, 2);
}

/**
 * 设置 upload_url_path，使用外部存储OSS
 *
 * @param $uploads
 * @return mixed
 */
function reset_upload_url_path( $uploads ) {
    $oss_options = get_option('oss_options', TRUE);

    if ($oss_options['static_url'] != "") {
        $baseurl = rtrim($oss_options['static_url'], '/');
        if(rtrim($oss_options['path'], '/') != ""){
            $baseurl = $baseurl .'/'. rtrim($oss_options['path'], '/');
        }
        $uploads['baseurl'] = $baseurl;
    }
    return $uploads;
}
add_filter( 'upload_dir', 'reset_upload_url_path', 30 );

/**
 * 添加设置页面入口连接
 *
 * @param $links
 * @param $file
 * @return array
 */
function oss_plugin_action_links( $links, $file ) {
    if ( $file == plugin_basename( dirname(__FILE__).'/oss-support.php' ) ) {
        $links[] = '<a href="options-general.php?page=' . OSS_BASEFOLDER . '/oss-support.php">'.__('Settings').'</a>';
    }
    return $links;
}
add_filter( 'plugin_action_links', 'oss_plugin_action_links', 10, 2 );

/**
 * 设置页面
 */
function oss_add_setting_page() {
    add_options_page('OSS Setting', 'OSS Setting', 'manage_options', __FILE__, 'oss_setting_page');
}
add_action('admin_menu', 'oss_add_setting_page');

function oss_setting_page() {
    
    $oss_options = get_option('oss_options');

    // Set(rewrite) the setting. 
    $options = array();
    if(isset($_POST['bucket'])) {
        $options['bucket'] = trim(stripslashes($_POST['bucket']));
    }
    if(isset($_POST['ak'])) {
        $options['ak'] = trim(stripslashes($_POST['ak']));
    }
    if(isset($_POST['end_point'])) {
        $options['end_point'] = trim(stripslashes($_POST['end_point']));
    }
    if(isset($_POST['sk'])) {
        if($_POST['sk'] === '') {
            $options['sk'] = $oss_options['sk'];
        } else {
            $options['sk'] = trim(stripslashes($_POST['sk']));
        }
    }
    if(isset($_POST['path'])) {
        $options['path'] = trim(trim(stripslashes($_POST['path'])), '/').'/';
    }
    if(isset($_POST['static_url'])) {
        $options['static_url'] = trim(stripslashes($_POST['static_url']));
    }
    if(isset($_POST['img_url'])) {
        $options['img_url'] = trim(stripslashes($_POST['img_url']));
    }
    if(isset($_POST['nolocalsaving'])) {
        $options['nolocalsaving'] = 'true';
    }

    if($options !== array() ){

        update_option('oss_options', $options);
        $oss_options = $options;
        ?>
        <div class="updated"><p><strong>设置已保存！</strong></p></div>
    <?php
    }

    // Show the setting. 
    $oss_bucket = isset($oss_options['bucket']) ? esc_attr($oss_options['bucket']) : null;
    $oss_ak = isset($oss_options['ak']) ? esc_attr($oss_options['ak']) : null;
    $oss_sk = isset($oss_options['sk']) ? esc_attr($oss_options['sk']) : null;
    $oss_path = isset($oss_options['path']) ? esc_attr($oss_options['path']) : null;
    $end_point = isset($oss_options['end_point']) ? esc_attr($oss_options['end_point']) : null;
    $oss_static_url = isset($oss_options['static_url']) ? esc_attr($oss_options['static_url']) : null;
    $oss_img_url = isset($oss_options['img_url']) ? esc_attr($oss_options['img_url']) : null;

    $oss_nolocalsaving = isset($oss_options['nolocalsaving']) ? esc_attr($oss_options['nolocalsaving']) : null;
    ($oss_nolocalsaving == 'true') ? ($oss_nolocalsaving = true) : ($oss_nolocalsaving = false);
    ?>
    <div class="wrap" style="margin: 10px;">
        <h2>阿里云存储 设置</h2>
        <form name="form1" method="post" action="<?php echo wp_nonce_url('./options-general.php?page=' . OSS_BASEFOLDER . '/oss-support.php'); ?>">
            <hr>
            <fieldset>
                <legend>Bucket 设置</legend>
                <input type="text" name="bucket" value="<?php echo $oss_bucket;?>" placeholder="请输入云存储使用的 bucket"/>
                <p>请先访问 <a href="http://i.aliyun.com/dashboard?type=oss">阿里云存储</a> 创建 bucket 后，填写以上内容。</p>
            </fieldset>
            <fieldset>
                <legend>Access Key / API key</legend>
                <input type="text" name="ak" value="<?php echo $oss_ak;?>" placeholder=""/>
            </fieldset>
            <br>
            <fieldset>
                <legend>Secret Key</legend>
                <?php if ( ! empty($oss_sk)) : ?>
                    <input type="text" name="sk" value="" placeholder="〄 你看不到我 ʅ(‾◡◝)"/> ←置空请填入空格
                <?php else : ?>
                    <input type="text" name="sk" value="<?php echo $oss_sk;?>" placeholder=""/>
                <?php endif ?>
                <p>访问 <a href="http://i.aliyun.com/access_key/" target="_blank">阿里云 密钥管理页面</a>，获取 AKSK</p>
            </fieldset>
            <fieldset>
                <legend>数据节点地址</legend>
                <input type="text" name="end_point" value="<?php echo $end_point;?>" placeholder=""/>
                <p>查看所有节点及地址 <a href="https://docs.aliyun.com/?spm=5176.7114037.1996646101.11.XMMlZa&pos=6#/pub/oss/product-documentation/domain-region" target="_blank">OSS数据中心地址</a></p>
            </fieldset>
            <hr>
            <fieldset>
                <legend>Save path on OSS</legend>
                <input type="text" name="path" value="<?php echo $oss_path;?>" placeholder="/"/>
                <P>可以设置在 OSS 上存储的位置，如果你要存在根目录就留空，我真的没意见(ﾉ￣д￣)ﾉ</P>
            </fieldset>
            <fieldset>
                <legend>OSS-Http-Url</legend>
                <input type="text" name="static_url" value="<?php echo $oss_static_url;?>" placeholder="http://"/>
                <P>OSS Bucket 的可访问 URL，支持已绑定到 OSS 的独立域名，留空将使用本地资源</P>
            </fieldset>
            <hr>
            <fieldset>
                <dl>
                    <dt>请瞩目：</dt>
                    <dd>1.图片服务是可选的，留空即可不启用</dd>
                    <dd>2.开启图片服务后，只有原图会上传到 OSS 中，缩略图不会再上传</dd>
                </dl>
                <legend>Aliyun-OSS 图片服务的 URL</legend>
                <input type="text" name="img_url" value="<?php echo $oss_img_url;?>" placeholder="http://"/>
            </fieldset>
            <hr>
            <fieldset>
                <label><input type="checkbox" name="nolocalsaving" <?php if($oss_nolocalsaving) echo 'checked="TRUE"';?> /> 不在本地保留备份</label>
            </fieldset>
            <fieldset class="submit">
                <input type="submit" name="submit" value="更新" />
            </fieldset>
        </form>
    </div>
<?php
}

/**
 * 删除本地的文件
 *
 * @param $file
 * @return bool
 */
function _delete_local_file($file){
    try{
        //文件不存在
        if(!@file_exists($file))
            return TRUE;
        //删除文件
        if(!@unlink($file))
            return FALSE;
        return TRUE;
    }
    catch(Exception $ex){
        return FALSE;
    }
}
